cdc/sink/mysql.go

```diff
@@ -389,163 +403,153 @@ func newMySQLSink(ctx context.Context, sinkURI *url.URL, dsn *dmysql.Config, fil
+	return &mysqlSinkWorker{
+		txnCh:     make(chan *model.Txn, 1024),
+		maxTxnRow: maxTxnRow,
+		bucket:    bucket,
+		execDMLs:  execDMLs,
+	}
+}
+
+func (w *mysqlSinkWorker) waitAllTxnsExecuted() {
+	w.txnWg.Wait()
 }
 
-func (w *mysqlSinkWorker) run(ctx context.Context) {
+func (w *mysqlSinkWorker) waitAndClose() {
+	w.waitAllTxnsExecuted()
+	close(w.txnCh)
+}
+
+func (w *mysqlSinkWorker) appendTxn(ctx context.Context, txn *model.Txn) {
+	if txn == nil {
+		return
+	}
+	w.txnWg.Add(1)
+	select {
+	case <-ctx.Done():
+		w.txnWg.Done()
+	case w.txnCh <- txn:
+	}
+}
+
+func (w *mysqlSinkWorker) run(ctx context.Context) (err error) {
 	defer func() {
 		if r := recover(); r != nil {
 			buf := make([]byte, 4096)
 			stackSize := runtime.Stack(buf, false)
 			buf = buf[:stackSize]
-			w.trySendErr(errors.Errorf("mysql sink concurrent execute panic, stack: %v", string(buf)))
+			err = errors.Errorf("mysql sink concurrent execute panic, stack: %v", string(buf))
 			log.Error("mysql sink worker panic", zap.Reflect("r", r), zap.Stack("stack trace"))
 		}
-		w.workerWg.Done()
 	}()
-	var multiRows [][]*model.RowChangedEvent
-	for rows := range w.rowsCh {
-		multiRows = append(multiRows, rows)
-		multiRows = w.fetchAllPendingEvents(multiRows)
-		for _, rows := range multiRows {
-			err := w.execDMLs(ctx, rows, w.bucket)
-			if err != nil {
-				w.trySendErr(err)
-			}
+
+	var toExecRows []*model.RowChangedEvent
+	var replicaID uint64
+	var txnNum int
+	lastExecTime := time.Now()
+	flushRows := func() error {
+		if len(toExecRows) == 0 {
+			return nil
 		}
-		// clean cache to avoid memory leak.
-		for i := range multiRows {
-			w.jobWg.Done()
-			multiRows[i] = nil
+		rows := make([]*model.RowChangedEvent, len(toExecRows))
+		copy(rows, toExecRows)
+		err := w.execDMLs(ctx, rows, replicaID, w.bucket)
+		if err != nil {
+			return err
 		}
-		multiRows = multiRows[:0]
+		toExecRows = toExecRows[:0]
+		w.txnWg.Add(-1 * txnNum)
+		txnNum = 0
+		lastExecTime = time.Now()
+		return nil
 	}
-}
 
-func (w *mysqlSinkWorker) trySendErr(err error) {
-	select {
-	case w.errCh <- err:
-	default:
-		return
-	}
-}
-
-func (w *mysqlSinkWorker) fetchAllPendingEvents(
-	multiRows [][]*model.RowChangedEvent,
-) [][]*model.RowChangedEvent {
-	// TODO(neil) count rows instead
-	for len(multiRows) < w.maxTxnRow {
+	t := time.NewTicker(50 * time.Millisecond)
+	defer t.Stop()
+	for {
 		select {
-		case rows, ok := <-w.rowsCh:
-			if !ok {
-				return multiRows
+		case <-ctx.Done():
+			return errors.Trace(flushRows())
+		case txn := <-w.txnCh:
+			if txn == nil {
+				return errors.Trace(flushRows())
+			}
+			if txn.ReplicaID != replicaID || len(toExecRows)+len(txn.Rows) > w.maxTxnRow {
+				if err := flushRows(); err != nil {
+					return errors.Trace(err)
+				}
+			}
+			replicaID = txn.ReplicaID
+			toExecRows = append(toExecRows, txn.Rows...)
+			txnNum++
+		case <-t.C:
+			if time.Since(lastExecTime) < time.Millisecond*100 {
```

Comment:
Why use another time gap checker in a time ticker

Reply:
I want to implement the following logic:

If there is no execution for more than 100 ms, then call `flushRows` 

but not:

Call `flushRows` every 100ms, whether or not it is triggered within this 100ms.